src/main/java/de/azapps/kafkabackup/common/partition/cloud/S3PartitionWriter.java

```diff
@@ -1,76 +1,104 @@
+  private Long lastCommittableOffset = null; // Nothing written so far, so nothing to commit
 
-  private final long lastWrittenOffset = 0;
 
+  public S3PartitionWriter(AwsS3Service awsS3Service, String bucketName, TopicPartition tp, int maxBatchMessages, long maxBatchTimeMs) {
+    this.awsS3Service = awsS3Service;
+    this.bucketName = bucketName;
+    this.topicPartition = tp;
+    this.maxBatchMessages = maxBatchMessages;
+    this.maxBatchTimeMs = maxBatchTimeMs;
+  }
 
   @Override
   public void append(Record record) throws PartitionException {
-    String fileName = buildFileKeyForRecord(topicName, partition, record);
-    try {
-      ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
-      recordJSONSerde.write(outputStream, record);
-      byte[] jsonRecord = outputStream.toByteArray();
-      ByteArrayInputStream jsonStream = new ByteArrayInputStream(jsonRecord);
+    buffer.add(record);
+  }
 
-      ObjectMetadata objectMetadata = new ObjectMetadata();
-      objectMetadata.setContentLength(jsonRecord.length);
+  @Override
+  public void flush() throws PartitionException, RetriableException {
+    while (!buffer.isEmpty()) {
+      Record record = buffer.poll();
+      writeToBatch(record);
+      maybeCommitBasedOnMessages();
+    }
+    maybeCommitBasedOnTime();
+  }
 
-      awsS3Service.saveFile(bucketName, fileName, jsonStream, objectMetadata);
+  private void writeToBatch(Record record) {
+    try {
+      if (batchWriter == null) {
+        batchWriter = new S3BatchWriter(awsS3Service, bucketName, topicPartition, record);
+      } else {
+        batchWriter.append(record);
+      }
     } catch (IOException e) {
-      e.printStackTrace();
+      throw new PartitionException(e);
     }
   }
 
-  private String buildFileKeyForRecord(String topic, int partition, Record record) {
-    return String.format("%s/%03d/msg_%020d.json", topic, partition, record.kafkaOffset());
+  private void maybeCommitBasedOnMessages() {
+    if (batchWriter != null && batchWriter.getCount() >= maxBatchMessages) {
+      log.info("Commit {} based on num messages", batchWriter.getObjectKey());
+      commitCurrentBatch();
+    }
   }
 
-  @Override
-  public void close() throws PartitionException {
-
+  private void maybeCommitBasedOnTime() {
+    long now = System.currentTimeMillis();
+    if (batchWriter != null && (now - batchWriter.getStartWallClockTime()) >= maxBatchTimeMs) {
+      log.info("Commit {} based on time", batchWriter.getObjectKey());
+      commitCurrentBatch();
+    }
   }
 
-  @Override
-  public void flush() throws PartitionException {
-
+  private void commitCurrentBatch() {
+    batchWriter.commitBatch();
+    // We got here without exception, so its safe to commit to kafka
+    lastCommittableOffset = batchWriter.getEndOffset() + 1;
+    batchWriter = null;
   }
 
   @Override
-  public long lastWrittenOffset() {
-    return lastWrittenOffset;
+  public void close() throws PartitionException {
```

Comment:
commitCurrentBatch shouldn't be called here? 🤔

Reply:
Hm... I'm not so sure. Note that when we get the call to close a partition from the connect framework, we are no longer responsible for this partition. So the result would be that we write some data to S3, but I would be surprised if we ever get a call to `preCommit` after that, so the write will not be committed back to the kafka cluster. Thus, any new task that get responsible for this partition will just write the data again. It seems safer to just close the partition and throw away any buffered records not yet written to S3 and hand it over to whoever takes over the partition.

Reply:
Note the overloading of the term "commit". `commitCurrentBatch` means writing to S3, and `preCommit` returns the offsets to commit to kafka.